### PR TITLE
[FLINK-26049][checkpoint] Improve logic after checkpoint trigger fails

### PR DESCRIPTION
**Pull request by 1996fanrui**
## What is the purpose of the change

Improve logic after checkpoint trigger fails

## Brief change log

1. show root cause if trigger checkpoint fails before creating PendingCheckpoint
2. Increase numberOfFailedCheckpoints if trigger checkpoint fails before creating PendingCheckpoint
3. initializeCheckpointLocation after create PendingCheckpoint

## Verifying this change

This change added tests and can be verified as follows:

  - *Add IOExceptionCheckpointIDCounter to check if the numberOfFailedCheckpoints can be increased when getCheckpointId throw IOException*
  - *Check if create PendingCheckpoint when initializeCheckpointLocation throw IOException*
  - *Check incrementFailedCheckpointsWithoutInProgress in CheckpointStatsCountsTest and CheckpointStatsTrackerTest*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not documented

### CODE REVIEW

**Comment by 1996fanrui** [comment]
@flinkbot run azure

**Comment by 1996fanrui** [comment]
@flinkbot run azure

**Comment by 1996fanrui** [comment]
@flinkbot run azure

**Comment by 1996fanrui** [comment]
@flinkbot run azure

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java [hunk 1]
```diff
@@ -931,11 +940,17 @@ private void onTriggerFailure(
...
+                        "Failed to trigger checkpoint for job {}. ({} consecutive failed attempts so far)",
                         job,
-                        throwable.getMessage());
+                        numUnsuccessful,
+                        throwable);
 
+                if (checkpoint == null
+                        && CheckpointFailureReason.IO_EXCEPTION
+                                == cause.getCheckpointFailureReason()) {
+                    statsTracker.reportFailedCheckpointsWithoutInProgress();
```
Comment (akalash) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:952]: I don't really agree with this condition. I mean if flink fails when the pending checkpoint has been created already then we update statistics independently on the reason for fail(even if the consecutive fail counter won't be increased). But if we have not created the pending checkpoint then we check for IO_EXCEPTION. In my opinion, it is not consistent behavior.

And in general, I don't really like how we use `CheckpointStatsCounts`. Right now we increase the checkpoint in progress count and the total checkpoint number only when the pending checkpoint has been created. But if we fail before the pending checkpoint we need to call the separated method(as it was created in this PR) to handle this situation. 
Maybe it is better to have logic like that:
- increment the number of inProgressCheckpoints and the total checkpoints as the first step before even checkpoint id has been created.
- add the PendingCheckpoint to stats tracker as soon as it would be created(right now it happens at the same time with previous step after the PendingCheckpoint is created)
- decrement the total number of checkpoints and checkpoints in progress as the last step of the checkpoint(the same as now)
- decrement checkpoints in progress but increment failed checkpoint number if the checkpoint fails(the same as now for PendingCheckpoint but use it for non existing PendingCheckpoint as well)

Important notice that the number of failed checkpoint in statistic and continuousFailureCounter in CheckpointFailureManager are totally different numbers. The first one can be grown all the time while the second one can be always zero if the reason of fail doesn't consider as critical.

@1996fanrui , @pnowojski , WDYT?

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/PendingCheckpoint.java [hunk 2]
```diff
@@ -105,12 +105,12 @@
...
     private final CheckpointProperties props;
 
-    /** Target storage location to persist the checkpoint metadata to. */
-    private final CheckpointStorageLocation targetLocation;
-
     /** The promise to fulfill once the checkpoint has been completed. */
     private final CompletableFuture<CompletedCheckpoint> onCompletionPromise;
 
+    /** Target storage location to persist the checkpoint metadata to. */
+    private CheckpointStorageLocation targetLocation;
```
Comment (akalash) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/PendingCheckpoint.java:112]: Since it is not final anymore we need to carefully think about the thread model here. I mean as I understand this field can be used from several threads. So perhaps it should be volatile or we should use it only under the lock.(I am not sure about that but it should be checked)

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java [hunk 3]
```diff
@@ -565,20 +563,37 @@ private void startTriggeringCheckpoint(CheckpointTriggerRequest request) {
...
+                                    },
+                                    executor)
+                            .thenComposeAsync(
+                                    (pendingCheckpoint) ->
+                                            OperatorCoordinatorCheckpoints
+                                                    .triggerAndAcknowledgeAllCoordinatorCheckpointsWithCompletion(
+                                                            coordinatorsToCheckpoint,
+                                                            pendingCheckpoint,
+                                                            timer),
+                                    timer);
```
Comment (akalash) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:596]: I think that it is better to extract changes of moving the checkpoint storage location to a separate commit(the same PR/branch but the second commit). I think it will be better since the logging/statistic fixes and these changes resolve the different problems. But it is better to fix only one problem in one commit.

**Review by akalash — COMMENTED** [review]
@1996fanrui , thanks a lot for these changes. I left a several comments in the PR, please, take a look. I also have an one question:
According to ticket, you had a problem with `flink won't execute the tolerable-failed-checkpoints logic. ` but as I see you didn't make any fix for that. Do I understand correctly, that you had such a problem because you have the old version of Flink but right now(with the master) it is not a problem anymore?

**Comment by 1996fanrui** [comment]
> @1996fanrui , thanks a lot for these changes. I left a several comments in the PR, please, take a look. I also have an one question: According to ticket, you had a problem with `flink won't execute the tolerable-failed-checkpoints logic. ` but as I see you didn't make any fix for that. Do I understand correctly, that you had such a problem because you have the old version of Flink but right now(with the master) it is not a problem anymore?

Hi @akalash , thanks for your review. 

Yeah, you are right. I have replied in JIRA that `our prod env use Flink 1.13. I see some jiras have resolved this issue.`, so I just fixed another problems.

https://issues.apache.org/jira/browse/FLINK-23189 and https://issues.apache.org/jira/browse/FLINK-24344

 I will address your comments and resubmit this PR as soon as possible. Thanks a lot.

**Comment by akalash** [comment]
@1996fanrui, we discussed with @pnowojski about this task a little. As result I have some important comments:

- I asked you moving logging from `CheckpointCoordinator` to `CheckpointFailureManager` but actually, it makes sense to do the same thing with `reportFailedCheckpoint`. So instead of reporting failed checkpoint from `PendingCheckpoint#abort` we can do it from `CheckpointFailureManager`.
- I also thought about your new method `reportFailedCheckpointsWithoutInProgress` perhaps you can leave it as is since my idea about calling the statistic before the checkpoint id increment is not so easy to implement.(especially the part with reporting failed checkpoint). 

As result, I think it makes sense to have three commits in this PR:

- Moving checkpoint location initialization(refactor commit)
- Moving logging and reporting failed checkpoint to `CheckpointFailureManager`(refactor commit)
- Changing the logging text and adding new logic about reporting checkpoint without pending checkpoint(bug fix commit)

Sorry, that the idea is slightly changed since my first comments but I think it will be easier to implement. And also if some of my proposals is not clear or not so good for you, please, give me now.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/PendingCheckpoint.java (hunk 2 above)
Comment (1996fanrui) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/PendingCheckpoint.java:112]: Thanks for your reminder. 
I will add volatile for targetLocation. And I think disposed also should be volatile, because there isn't use lock in `isDisposed` method. How do you think?

**Comment by 1996fanrui** [comment]
> @1996fanrui, we discussed with @pnowojski about this task a little. As result I have some important comments:
> 
>     * I asked you moving logging from `CheckpointCoordinator` to `CheckpointFailureManager` but actually, it makes sense to do the same thing with `reportFailedCheckpoint`. So instead of reporting failed checkpoint from `PendingCheckpoint#abort` we can do it from `CheckpointFailureManager`.
> 
>     * I also thought about your new method `reportFailedCheckpointsWithoutInProgress` perhaps you can leave it as is since my idea about calling the statistic before the checkpoint id increment is not so easy to implement.(especially the part with reporting failed checkpoint).
> 
> 
> As result, I think it makes sense to have three commits in this PR:
> 
>     * Moving checkpoint location initialization(refactor commit)
> 
>     * Moving logging and reporting failed checkpoint to `CheckpointFailureManager`(refactor commit)
> 
>     * Changing the logging text and adding new logic about reporting checkpoint without pending checkpoint(bug fix commit)
> 
> 
> Sorry, that the idea is slightly changed since my first comments but I think it will be easier to implement. And also if some of my proposals is not clear or not so good for you, please, give me now.

Hi @akalash , I have resubmitted this PR with 3 commits. Please help to review again in your free time, thanks a lot.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java [hunk 4]
```diff
@@ -565,20 +563,37 @@ private void startTriggeringCheckpoint(CheckpointTriggerRequest request) {
...
+                            .thenApplyAsync(
+                                    pendingCheckpoint -> {
+                                        try {
+                                            CheckpointStorageLocation checkpointStorageLocation =
+                                                    initializeCheckpointLocation(
+                                                            pendingCheckpoint.getCheckpointID(),
+                                                            request.props,
+                                                            request.externalSavepointLocation,
+                                                            initializeBaseLocations);
+                                            pendingCheckpoint.setCheckpointTargetLocation(
```
Comment (akalash) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:577]: As I see now, according to our thread model here. It is better to avoid the volatile inside pendingCheckpoint. You can move setting the StorageLocation under the lock and to `timer` thread. So in this case you can move `pendingCheckpoint.setCheckpointTargetLocation` to next `thenComposeAsync` or create new small one. And use `synchronization(lock)`around the setting targetLocation. And then you can remove volatile for both fields(targetLocation, disposed).
I know it is not fully obvious but in parallel to your PR there is the task for moving all calls to JobManagerThread that make all execution here one threaded and it will allow avoiding synchronization at all. It is why it doesn't make sense to add new volatile in this PR.

**Review by akalash — COMMENTED** [review]
@1996fanrui , thanks for updating the PR. It looks almost done for me. I have left only a couple of comments in PR.

**Comment by 1996fanrui** [comment]
> 

Hi @akalash , thank you again for your review. I have solved your all comments.

**Comment by 1996fanrui** [comment]
> @1996fanrui , thanks for updating the PR. It looks almost done for me. I have left only a couple of comments in PR.

Hi @akalash , thank you again for your review. I have solved your all comments.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java [hunk 5]
```diff
@@ -565,20 +563,44 @@ private void startTriggeringCheckpoint(CheckpointTriggerRequest request) {
...
+                                                    pendingCheckpoint, checkpointStorageLocation);
+                                        } catch (Throwable e) {
+                                            throw new CompletionException(e);
+                                        }
+                                    },
+                                    executor)
+                            .thenApplyAsync(
+                                    (checkpointInfo) -> {
+                                        PendingCheckpoint pendingCheckpoint = checkpointInfo.f0;
+                                        pendingCheckpoint.setCheckpointTargetLocation(
```
Comment (akalash) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:587]: Sorry, perhaps, I wasn't clear enough. But I meant that you should use the lock here rather than inside of pendingCheckpoint. So please remove synchronization from `setCheckpointTargetLocation` and add it here:
```
synchronization(lock) {
  pendingCheckpoint.setCheckpointTargetLocation(checkpointInfo.f1);
}
```
We need it because as you can see `targetLocation` in `PendingCheckpoint#discard` is used outside of PendingCheckpoint#lock but it is always used with CheckpointCoordinator#lock or in `timer` thread.

**Review by akalash — COMMENTED** [review]
@1996fanrui, thanks again. I have left one small comment. I believe it will be last.

**Comment by 1996fanrui** [comment]
> @1996fanrui, thanks again. I have left one small comment. I believe it will be last.

@akalash , thanks again. I have updated.

**Review by akalash — APPROVED** [review]
LGTM. let's wait for the green azure and after that @pnowojski can you help with the merge, please?

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java [hunk 6]
```diff
@@ -545,14 +545,12 @@ private void startTriggeringCheckpoint(CheckpointTriggerRequest request) {
...
                                     plan -> {
                                         try {
-                                            CheckpointIdAndStorageLocation
-                                                    checkpointIdAndStorageLocation =
-                                                            initializeCheckpoint(
-                                                                    request.props,
-                                                                    request.externalSavepointLocation,
-                                                                    initializeBaseLocations);
-                                            return new Tuple2<>(
-                                                    plan, checkpointIdAndStorageLocation);
```
[the remark this answers is not in the document]
Comment (pnowojski) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:555]: Also I have some small doubts that we are splitting one async IO action into two smaller ones, which can cause some extra overheads. As I understand the reason is that you want to have an existing pending checkpoint when doing error handling of `initializeCheckpointLocation`, right? Wouldn't an alternate solution to this problem be to have somewhere in the error handling logic, if we are handling an error for a checkpoint that's missing `PendingCheckpoint`, we lazily/on-demand create one? Have you considered this approach and what pros/cons would it have?

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/PendingCheckpoint.java (hunk 2 above)
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/PendingCheckpoint.java:112]: nit: `@Nullable`

**Comment by pnowojski** [comment]
Thanks @1996fanrui and @akalash for the PR and reviewing. Before merging I had a couple of questions.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (1996fanrui) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:555]: > initializing the storage location was moved after creating pending checkpoint due to error handling reasons, right? If so, why leaving `checkpointIdCounter.getAndIncrement()` here doesn't cause the same problems?

We have handled checkpoint error, we `initialize CheckpointLocation after create PendingCheckpoint` due to we want to PendingCheckpoint as early as possible. It will show Checkpoint in CheckpointHistory Page. But we must leave  `checkpointIdCounter.getAndIncrement()` here, because checkpoint is necessary for create PendingCheckpoint.

If checkpoint throw exception before create pendingCheckpoint, it won't show checkpoint info in CheckpointHistory Page.

[the remark this answers is not in the document]
Comment (akalash) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:555]: > initializing the storage location was moved after creating pending checkpoint due to error handling reasons, right? If so, why leaving checkpointIdCounter.getAndIncrement() here doesn't cause the same problems?

Yes, it is correct. It will be the same problem for getAndIncrement but these changes just try to reduce the probability of this event, and also if we don't even have the checkpoint id then there is nothing to send to the statistic.

> Wouldn't an alternate solution to this problem be to have somewhere in the error handling logic, if we are handling an error for a checkpoint that's missing PendingCheckpoint, we lazily/on-demand create one?

I thought about that but I didn't find an easy way to do so. I mean right now we lose all information if we fail before the PendingCheckpoint. I also think that PendingCheckpoint is too complex an object for such fake creation. I am maybe wrong but it looks like it requires a lot of refactoring. Considering that this ticket is about logging, statistic but not about PendingCheckpoint lifecycle, I decided that we can do more deep refactoring in the other ticket if we think that it makes sense.

[the remark this answers is not in the document]
Comment (akalash) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:555]: @1996fanrui, maybe you have something to add to my answers?

[the remark this answers is not in the document]
Comment (1996fanrui) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:555]: Hi @akalash your answer is comprehensive, I agree with do more deep refactoring in the other ticket.

[the remark this answers is not in the document]
Comment (pnowojski) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:555]: Ok, thanks for the clarifications.
